cabal-install/main/Main.hs

```diff
@@ -332,39 +332,74 @@ wrapperAction command verbosityFlag distPrefFlag =
     setupWrapper verbosity setupScriptOptions Nothing
                  command (const flags) extraArgs
 
+type ProgramPaths = [(String, FilePath)]
+type ProgramArgs  = [(String, [String])]
+
+addTargetFlags :: Flag String   -- ^ target
+               -> ProgramPaths
+               -> ProgramArgs
+               -> Flag FilePath -- ^ dist pref
+               -> [String]      -- ^ configure args
+               -> (ProgramPaths, ProgramArgs, Flag FilePath, [String])
+addTargetFlags NoFlag p a d c = (p, a, d, c)
+addTargetFlags (Flag target) p a d c = (p', a', d', c')
+  where p' =  ("ghc",     fromMaybe (target ++ "-ghc")     (lookup "ghc" p))
+             :("ghc-pkg", fromMaybe (target ++ "-ghc-pkg") (lookup "ghc-pkg" p))
+             :filter (not . flip elem ["ghc", "ghc-pkg"] . fst) p
+        a' =  ("hsc2hs",   "--cross-compile":fromMaybe [] (lookup "hsc2hs" a))
+             :filter (not . flip elem ["hsc2hs"] . fst) a
+        d' = toFlag ("dist" </> target) <> d
+        c' = ("--host="++target):c
+
 configureAction :: (ConfigFlags, ConfigExFlags)
                 -> [String] -> Action
 configureAction (configFlags, configExFlags) extraArgs globalFlags = do
-  let verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
-  (useSandbox, config) <- updateInstallDirs (configUserInstall configFlags)
+
+  let verbosity = fromFlagOrDefault normal (configVerbosity configFlags')
```

Comment:
NB: This modification (and others) apply to the *legacy* cabal-install codepaths `cabal configure` and `cabal build`. Thus, if you use `cabal new-build`, the modifications you've applied here won't apply at all.